### ISSUE TITLE
Git clone recursive with keys hangs 

### ISSUE THREAD

**Issue by fbrbovic** [ansible/ansible-modules-core#1343]
Issue Type:

Bug Report.

Ansible Version:

Ansible 1.9.1

Ansible Configuration:

No Changes

Environment:

Centos 7 with tower installed

Summary:

When cloning a repo which has submodules and using an ssh key ansible hangs on the submodule part. 
Once I stop the run and log on to the server I see that initial Repo is checked out but not the submodules. Also doing `ps aux | grep git` I see that git is just hanging there with following processes. 

```
root     22124  0.0  0.0  12852   940 pts/1    S+   May15   0:00 /bin/git submodule update --init --recursive
root     22125  0.0  0.0   9772  1768 pts/1    S+   May15   0:00 /bin/sh /usr/libexec/git-core/git-submodule update --init --recursive
root     22187  0.0  0.0   9772  1324 pts/1    S+   May15   0:00 /bin/sh /usr/libexec/git-core/git-submodule update --init --recursive
root     22289  0.0  0.0   9900  1076 pts/1    S+   May15   0:00 /bin/sh /usr/libexec/git-core/git-submodule update --init --recursive
root     22291  0.0  0.0  12860  1248 pts/1    S+   May15   0:00 git clone -n --separate-git-dir .git/modules/core https://github.com/CHANGED_REPO_NAME
root     22292  0.0  0.3 173124  7228 pts/1    S+   May15   0:00 git-remote-https origin https://github.com/CHANGED_REPO_NAME
```

It looks like the passed ssh key is not being used during git submodule update. 

This is the role I have : 

```

---
- name: create app deploy ~/.ssh directory
  file: path=/root/.ssh
    state=directory
    mode=0700
    owner=root
    group=root
  tags: ['deploy-ssh-keys']

- name: Create the github SSH public key file
  copy: src={{ git_public_key }}
        dest=/root/.ssh/id_rsa_github.pub
        mode=0644
        owner=root
        group=root
  tags: ['deploy-ssh-keys']

- name: Create the github SSH private key file
  copy: src={{ git_private_key }}
        dest=/root/.ssh/id_rsa_github
        mode=0600
        owner=root
        group=root
  tags: ['deploy-ssh-keys']

- name: ensure github.com is a known host
  lineinfile:
    dest: /root/.ssh/known_hosts
    create: yes
    state: present
    line: "{{ lookup('pipe', 'ssh-keyscan -t rsa github.com') }}"
    regexp: "^github\\.com"

- name: Clone project files 
  git: repo={{ git_repo }} dest={{ root_path }}/release/{{ version }} version={{ version }} depth=1 accept_hostkey=yes key_file=/root/.ssh/id_rsa_github force=yes
```

Any idea how to fix this ?

**Comment by bennojoy** [comment]
@fbrbovic  i am not able to reproduce this with 1.9.1 or the latest devel branch, can you please try manually cloning the repo with the key and user that is used in the playbook and check if that works.

**Comment by twneale** [comment]
I'm having the same issue with ansible-playbook 1.8.4 on RHEL 6.4, only without even the recursive or ssh key options:

```
- name: mmlib.update
  git:
    dest: "{{mmlib_home}}/{{repo}}"
    accept_hostkey: yes
    name: "git://{{git_host}}/{{repo}}.git"
    force: yes
```

This one hangs forever after printing:

failed: [devlx105.private.cows.com] => {"failed": true}
msg: Unable to determine hashes of submodules

**Comment by taraspos** [comment]
Ansible  1.9.1
Centos 6

I'm facing the same issue

```
- name: clone files from repo
  git:
    repo: 'git@bitbucket.org:{{website.repo}}.git'
    dest: '{{ website.root }}/{{ website.name }}'
    accept_hostkey: 'yes'
    force: 'yes'
    key_file: '{{ website.keys_location }}/id_rsa'
  remote_user: '{{ website.user }}'
```

when I'm logging in into server, going into `{{ website.root }}` directory and running `git clone git@bitbucket.org:{{website.repo}}.git'` I'm getting `Accept Hostkey` request and downloading starts normally. 

if i run playbook again, everything is fine. Issue exist only at the initial checkout.

Also, in the next role, another project should be cloned from BitBucket, so, second project  is being cloned well, i believe this is so, because BitBucket's hostkey already Accepted

**Comment by twneale** [comment]
I was able to confirm that in my case the failure was do to shenanigans related to submodule file and/or repo permissions. You can use pdb to get the output of the git commands ansible runs when it checks the submodule hashes, and in my case it was obvious that various permissions were thoroughly hosed up.

**Comment by index0h** [comment]
Hi, I'm have same issue
OS: Centos 6.4
Ansible: 1.9.2

``` yml
- name: clone files from repo
  git: repo='ssh://my_user@my_host:MY_PORT/my_repo.git'
    dest: '{{ website.root }}/{{ website.name }}'
    accept_hostkey: 'yes'
  sudo: no
```

In case of not default port:

``` sh
ssh-key -t rsa my_host
```

will receive empty answer -> pub key will not be stored in known_hosts

**Comment by galindro** [comment]
Guys,

This problem seems to be old, like reported in https://github.com/ansible/ansible/issues/7474 and by me yesterday in https://github.com/ansible/ansible/issues/12840. It isn't related to download submodules from git, but I think it is related to create the know_hosts file in home user.

My case is: My git task hangs because Ansible's git module ignores the parameter accept_hostkey: true. This is my task:

``` yaml
- name: Pull sources from the repository
  git: 
    repo: "{{ app_repository }}"
    dest: "{{ app_codebase }}"
    version: "{{ app_version }}"
    accept_hostkey: "true"
    force: "true"
    key_file: "{{ ssh_key_private }}"
  sudo: yes
  notify:
    - restart php5-fpm
    - change user and group on app_codebase
    - change file permissions on app_codebase
  tags:
    - app
    - sb_upgrade
```

I've executed my playbook with -vvvv and this line is printed before ansible hangs. After it, no more actions are performed. This is because, in the host, it is waiting for user to anwser yes to this question: 

``` bash
The authenticity of host 'bitbucket.org (131.103.20.167)' can't be established.
RSA key fingerprint is 97:8c:1b:f2:6f:14:6b:5c:3b:ec:aa:46:46:74:7c:40.
Are you sure you want to continue connecting (yes/no)? 
```

``` bash
ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/var/lib/jenkins/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o Port=40 -o IdentityFile="app.pem" -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 54.172.38.244 /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=kenuyfqnmeunvwzzxwlflrevqsgwphjx] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-kenuyfqnmeunvwzzxwlflrevqsgwphjx; LANG=en_US.UTF-8 TMOUT=600 LC_CTYPE=en_US.UTF-8 /usr/bin/python /home/ubuntu/.ansible/tmp/ansible-tmp-1445353624.41-274068595469193/git; rm -rf /home/ubuntu/.ansible/tmp/ansible-tmp-1445353624.41-274068595469193/ >/dev/null 2>&1'"'"''
```

After that problem, I've logged into the server and copied this python file: /home/ubuntu/.ansible/tmp/ansible-tmp-1445353624.41-274068595469193/git to my workstation. I've started a new empty virtual machine and copied that python file to it. So, I've executed this commands to reproduce the error:

``` bash
ubuntu@ip-172-30-2-232:~$ sudo -k
ubuntu@ip-172-30-2-232:~$ sudo -H -S -p "[sudo via ansible, key=kenuyfqnmeunvwzzxwlflrevqsgwphjx] password: " -u root /bin/sh -c "echo BECOME-SUCCESS-kenuyfqnmeunvwzzxwlflrevqsgwphjx; LANG=en_US.UTF-8 TMOUT=600 LC_CTYPE=en_US.UTF-8 /usr/bin/python -m trace --trace git > git.trace"
BECOME-SUCCESS-kenuyfqnmeunvwzzxwlflrevqsgwphjx

The authenticity of host 'bitbucket.org (131.103.20.167)' can't be established.
RSA key fingerprint is 97:8c:1b:f2:6f:14:6b:5c:3b:ec:aa:46:46:74:7c:40.
Are you sure you want to continue connecting (yes/no)? ^Cubuntu@ip-172-30-2-232:~$ 
```

As we can see above, I could reproduced it with trace info saved in git.trace. You could see those files here (I've changed the bitbucket url on booth files):

[git](https://gist.github.com/galindro/04e28437c7ffcf46e4fa)
[git.trace](https://gist.github.com/galindro/980fb960fdd6f27396ec)

The above command created a know_hosts file in /root/.ssh/known_hosts with this content. It think it is normal because I've run /usr/bin/python with trace...

``` bash
root@ip-172-30-2-232:~# cat .ssh/known_hosts 
cess.py(1150):                 cloexec_flag = fcntl.FD_CLOEXEC
subprocess.py(1154):             old = fcntl.fcntl(fd, fcntl.F_GETFD)
subprocess.py(1155):             if cloexec:
subprocess.py(1156):                 fcntl.fcntl(fd, fcntl.F_SETFD, old | cloexec_flag)
subprocess.py(1169):             self._set_cloexec_flag(w)
 --- modulename: subprocess, funcname: _set_cloexec_flag
subprocess.py(1149):             try:
subprocess.py(1150):                 cloexec_flag = fcntl.FD_CLOEXEC
subprocess.py(1154):             old = fcntl.fcntl(fd, fcntl.F_GETFD)
subprocess.py(1155):             if cloexec:
subprocess.py(1156):                 fcntl.fcntl(fd, fcntl.F_SETFD, old | cloexec_flag)
subprocess.py(1170):             return r, w
subprocess.py(1216):             try:
subprocess.py(1217):                 try:
subprocess.py(1218):                     gc_was_enabled = gc.isenabled()
subprocess.py(1221):                     gc.disable()
subprocess.py(1222):                     try:
subprocess.py(1223):                         self.pid = os.fork()
subprocess.py(1228):                     self._child_created = True
subprocess.py(1229):                     if self.pid == 0:
subprocess.py(1231):                         try:
subprocess.py(1233):                             if p2cwrite is not None:
subprocess.py(1235):                             if c2pread is not None:
subprocess.py(1236):                                 os.close(c2pread)
subprocess.py(1237):                             if errread is not None:
subprocess.py(1238):                                 os.close(errread)
subprocess.py(1239):                             os.close(errpipe_read)
subprocess.py(1244):                             if c2pwrite == 0:
subprocess.py(1246):                             if errwrite == 0 or errwrite == 1:
subprocess.py(1250):                             def _dup2(a, b):
subprocess.py(1258):                             _dup2(p2cread, 0)
 --- modulename: subprocess, funcname: _dup2
subprocess.py(1254):                                 if a == b:
subprocess.py(1256):                                 elif a is not None:
subprocess.py(1259):                             _dup2(c2pwrite, 1)
 --- modulename: subprocess, funcname: _dup2
subprocess.py(1254):                                 if a == b:
subprocess.py(1256):                                 elif a is not None:
subprocess.py(1257):                                     os.dup2(a, b)
subprocess.py(1260):                             _dup2(errwrite, 2)
 --- modulename: subprocess, funcname: _dup2
subprocess.py(1254):                                 if a == b:
subprocess.py(1256):                                 elif a is not None:
subprocess.py(1257):                                     os.dup2(a, b)
subprocess.py(1264):                             closed = { None }
subprocess.py(1265):                             for fd in [p2cread, c2pwrite, errwrite]:
subprocess.py(1266):                                 if fd not in closed and fd > 2:
subprocess.py(1265):                             for fd in [p2cread, c2pwrite, errwrite]:
subprocess.py(1266):                                 if fd not in closed and fd > 2:
subprocess.py(1267):                                     os.close(fd)
subprocess.py(1268):                                     closed.add(fd)
subprocess.py(1265):                             for fd in [p2cread, c2pwrite, errwrite]:
subprocess.py(1266):                                 if fd not in closed and fd > 2:
subprocess.py(1267):                                     os.close(fd)
subprocess.py(1268):                                     closed.add(fd)
subprocess.py(1265):                             for fd in [p2cread, c2pwrite, errwrite]:
subprocess.py(1270):                             if cwd is not None:
subprocess.py(1273):                             if preexec_fn:
subprocess.py(1278):                             if close_fds:
subprocess.py(1279):                                 self._close_fds(but=errpipe_write)
 --- modulename: subprocess, funcname: _close_fds
subprocess.py(1174): 
```

After executed the above commands, I've deleted /root/.ssh/knows_hosts file and executed the script again from ubuntu account wihout trace information and with a simple sudo command, and It worked without errors.

``` bash
ubuntu@ip-172-30-2-232:~$ sudo /usr/bin/python git
{"changed": true, "after": "8fdad57ea5e7dcdc8bc72075f1536bdcb98f4ade", "before": null}
```

So, I've deleted again my /root/.ssh/known_hosts file, cleared /var/www/app local repository and executed /usr/bin/python git it with the same commands that I've posted first, but now without trace information. And the problem is showed again. The git module created an empty known_hosts file because I didn't accepted it (I've terminated the script with Ctrl+C like the first one executed with trace).

``` bash
ubuntu@ip-172-30-2-232:~$ sudo -k
ubuntu@ip-172-30-2-232:~$ sudo -H -S -p "[sudo via ansible, key=kenuyfqnmeunvwzzxwlflrevqsgwphjx] password: " -u root /bin/sh -c "echo BECOME-SUCCESS-kenuyfqnmeunvwzzxwlflrevqsgwphjx; LANG=en_US.UTF-8 TMOUT=600 LC_CTYPE=en_US.UTF-8 /usr/bin/python git"
BECOME-SUCCESS-kenuyfqnmeunvwzzxwlflrevqsgwphjx
The authenticity of host 'bitbucket.org (131.103.20.167)' can't be established.
RSA key fingerprint is 97:8c:1b:f2:6f:14:6b:5c:3b:ec:aa:46:46:74:7c:40.
Are you sure you want to continue connecting (yes/no)? ^C{"msg": "Traceback (most recent call last):\n  File \"git\", line 2302, in run_command\n    rfd, wfd, efd = select.select(rpipes, [], rpipes, 1)\nKeyboardInterrupt\n", "failed": true, "cmd": "/usr/bin/git ls-remote '' -h refs/heads/master", "rc": 257}
ubuntu@ip-172-30-2-232:~$ sudo ls -lh /root/.ssh/
total 12K
-rw------- 1 root root  555 Oct 21 09:10 authorized_keys
-rw------- 1 root root 1.7K Aug 13 15:52 id_rsa
-rw-r--r-- 1 root root  381 Aug 13 15:52 id_rsa.pub
-rw-r--r-- 1 root root    0 Oct 21 10:42 known_hosts
```

**Comment by galindro** [comment]
If I execute a task that adds the bitbucket key to known_hosts before execute the git task, the issue doesn't occurs, like proposed in https://github.com/ansible/ansible/issues/7474

``` yaml
- name: Add git_server key to known_hosts
  shell: ssh-keyscan -H -t dsa {{ git_server }} >> ~/.ssh/known_hosts 2>&1
  sudo: yes
  tags:
    - ssh
```

**Comment by fastinetserver** [comment]
Same problem on ansible 1.9.4 (under ubuntu-desktop 12.04 64 bit).
Task to reproduce:

```
- name: clone the site from git repo
  git: repo=gl:{{project_name}} dest="/home/someuser/git/{{project_name}}" accept_hostkey=yes key_file=~/.ssh/private_key
```

So i use this workaround:

```
- name: Add gl to known_hosts (workaround to avoid frozen git clone task)
  command: ssh -oStrictHostKeyChecking=no gl
- name: clone the site from git repo
  git: repo=gl:{{project_name}} dest="/home/someuser/git/{{project_name}}" accept_hostkey=yes key_file=~/.ssh/private_key
```

**Comment by galindro** [comment]
Guys any return about this problem?

**Comment by gretel** [comment]
still running into this issue using `2.0.1.0` - please assist.

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by galindro** [comment]
Guys any return about this problem?

**Comment by sean-abbott** [comment]
Still having this issue in 2.1.1.0.

**Comment by sean-abbott** [comment]
Ok, just in case anyone else has this problem I don't know that this is what I was running into.  I was running ansible with deletegate_to and forward agent, but apparently become was messing with forward agent. I was able to make everything work by using "become: no", as per this suggestion: http://stackoverflow.com/questions/20952689/ansible-ssh-forwarding-doesnt-seem-to-work-with-vagrant

And magically git started working again.

Apparently sudoing to root and back to the ssh user seems like a surprising principle.

**Comment by jctanner** [comment]
I am unable to reproduce this issue with the following playbook on any ansible version since 1.7.2 .

```
- hosts: localhost
  connection: local
  gather_facts: False
  vars:
    git_repo: ssh://git@github.com/jctanner/ansible
    version: stable-2.0
    dest: /tmp/ansible.submodule.check
  tasks:
    - file:
        path: "{{ dest }}"
        state: absent
    - name: Clone project files
      git:
        repo: "{{ git_repo }}"
        dest: "{{ dest }}"
        version: "{{ version }}"
        depth: 1
        accept_hostkey: yes
        key_file: ~/.ssh/id_rsa
        force: yes
```

Since there has been no additional updates from the issue submitter and each subsequent response pertains to slightly different scenarios, I'm going to close this. If you have a similar issue, please provide a playbook and -vvvv output in a new ticket.
